samples/java-protobuf-tracing/src/main/java/com/example/ControllerAction.java

```diff
@@ -0,0 +1,87 @@
+package com.example;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Charsets;
+import com.google.protobuf.Empty;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Scope;
+import kalix.javasdk.action.ActionCreationContext;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.Charset;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public class ControllerAction extends AbstractControllerAction {
+
+  String url = "https://jsonplaceholder.typicode.com/posts/1";
+  HttpClient httpClient = HttpClient.newHttpClient();
+  public ControllerAction(ActionCreationContext creationContext) {}
+
+
+  @Override
+  public Effect<ControllerActionApi.MessageResponse> callAsyncEndpoint(Empty empty) {
+    // tag::get-tracer[]
+    Optional<Tracer> tracerOpt = actionContext().getOpenTelemetryTracer();
+    // end::get-tracer[]
+    CompletableFuture<HttpResponse<Post>> futureResponse;
+
+    if(tracerOpt.isPresent()) {
+      futureResponse = callAsyncService(tracerOpt.get());
```

Comment:
I know it's a matter of taste, but can't we just include the `CompletableFuture<HttpResponse<Post>> callAsyncService(Tracer tracer)` code here and remove `futureResponse`.  Instead of using `Optional<Trace>` we can use `Optional<Span>` as suggested in a different command and be more FP. WDYT?

Reply:
A link: https://github.com/lightbend/kalix-jvm-sdk/pull/2051#discussion_r1538939313 creating the span might be in the same method.

Reply:
Certainly is more elegant the way you propose. I like it very much. At least for a span without attributes. What I'm no so sure is how it will look when starting to add the attributes to the span. One `ifPresent` for each? On that note, I had a discussion with @octonato yesterday about providing a Tracer that can be a NoOpsTracer when tracing is disabled, so the user doesn't have to deal with all this mess of having two flows. `IfPresent` namely in any of the two flavours, the current one or the more functional you propose.
All that to say that I rather leave it the way it is until I add the NoOps Tracer.

Reply:
No, you need a single if present, here my impl:
```
  @Override
  public Effect<ControllerActionApi.MessageResponse> callAsyncEndpoint(Empty empty) {
    // tag::get-tracer[]
    Optional<Tracer> tracerOpt = actionContext().getOpenTelemetryTracer();
    // end::get-tracer[]

    Optional<Span> span = tracerOpt.map(tracer -> {
      return tracer
        .spanBuilder("https://jsonplaceholder.typicode.com/posts/{}")
        .setParent(actionContext().metadata().traceContext().asOpenTelemetryContext())// <1>
        .startSpan() // <2>
        .setAttribute("post", "1");// <3>
    });

    CompletableFuture<ControllerActionApi.MessageResponse> asyncComputation = callAsyncService()
      .whenComplete((response, ex) -> {
        if (ex != null) {
          span.ifPresent(presentSpan ->
            presentSpan
              .setStatus(StatusCode.ERROR, ex.getMessage())// <6>
              .end());// <5>
        } else {
          span.ifPresent(presentSpan ->
            presentSpan
              .setAttribute("result", response.body().title)// <4>
              .end());// <5>
        }
      })
      .thenApply(response ->
        ControllerActionApi.MessageResponse.newBuilder().setMessage(response.body().title).build()
      );
    return effects().asyncReply(asyncComputation);
  }
```

Java Futures API is awful, but I think in this case `whenComplete` will do the job better.

I also like the idea of returning `NoOps Tracer` but this shouldn't block us with decisions about the current implementation.

Reply:
I can't say no with this beautiful snippet. I'll put it in. 👍 

PS. Two `.ifPresent` :)